src/number.rs

```diff
@@ -292,90 +390,412 @@ impl<'de> Deserialize<'de> for Number {
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string containing a number")
+            }
+
+            fn visit_string<E>(self, s: String) -> Result<NumberFromString, E>
+            where E: de::Error,
+            {
+                Ok(NumberFromString { value: Number::from_string_unchecked(s) })
+            }
+        }
+
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+fn invalid_number() -> Error {
+    Error::syntax(ErrorCode::InvalidNumber, 0, 0)
+}
+
+macro_rules! deserialize_number {
+    ($deserialize:ident => $visit:ident) => {
+        #[cfg(not(feature = "arbitrary_precision"))]
+        fn $deserialize<V>(self, visitor: V) -> Result<V::Value, Error>
+        where
+            V: Visitor<'de>,
+        {
+            self.deserialize_any(visitor)
+        }
+
+        #[cfg(feature = "arbitrary_precision")]
+        fn $deserialize<V>(self, visitor: V) -> Result<V::Value, Error>
+        where
+            V: de::Visitor<'de>,
+        {
+            visitor.$visit(self.n.parse().map_err(|_| invalid_number())?)
+        }
+    }
+}
+
 impl<'de> Deserializer<'de> for Number {
     type Error = Error;
 
+    #[cfg(not(feature = "arbitrary_precision"))]
     #[inline]
     fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
     where
         V: Visitor<'de>,
     {
         match self.n {
-            N::PosInt(i) => visitor.visit_u64(i),
+            N::PosInt(u) => visitor.visit_u64(u),
             N::NegInt(i) => visitor.visit_i64(i),
             N::Float(f) => visitor.visit_f64(f),
         }
     }
 
+    #[cfg(feature = "arbitrary_precision")]
+    #[inline]
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
+        where V: Visitor<'de>
+    {
+        visitor.visit_map(NumberDeserializer {
+            visited: false,
+            number: self.n.into(),
+        })
+    }
+
+    deserialize_number!(deserialize_i8 => visit_i8);
+    deserialize_number!(deserialize_i16 => visit_i16);
+    deserialize_number!(deserialize_i32 => visit_i32);
+    deserialize_number!(deserialize_i64 => visit_i64);
+    deserialize_number!(deserialize_u8 => visit_u8);
+    deserialize_number!(deserialize_u16 => visit_u16);
+    deserialize_number!(deserialize_u32 => visit_u32);
+    deserialize_number!(deserialize_u64 => visit_u64);
+    deserialize_number!(deserialize_f32 => visit_f32);
+    deserialize_number!(deserialize_f64 => visit_f64);
+
+    #[cfg(not(feature = "arbitrary_precision"))]
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: Visitor<'de>,
+    {
+        self.deserialize_any(visitor)
+    }
+
+    #[cfg(feature = "arbitrary_precision")]
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_str(&self.n)
```

Comment:
Is this an intentional behavior change or required somewhere else in the PR? The following fails without `arbitrary_precision` but gives you a string with `arbitrary_precision`.

```rust
println!("{:?}", String::deserialize(Number::from(0)).unwrap());
```

We have tended to stay away from such implicit type conversions because they make other functionality fail in surprising ways, for example deserializing to the following untagged enum would produce the string variant even though the input is a number.

```rust
#[derive(Deserialize)]
#[serde(untagged)]
enum E {
    S(String),
    N(i64),
}
```

Reply:
Ah. It was so long ago I originally wrote this, I don't fully remember my motivations, but I think it was mainly for convenience. Your point is well taken though.

Reply:
Please remove the methods that are resulting in a behavior change.